text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+the book. This may be an adequate workaround for functions, but there is
+no way to extend this to work with impls or traits. As a result, the
+following still hold:
+
+ - Implementing traits for `Tuple` could use some cleanup. For example, it
+   is currently impossible to compare two `Tuple`s with `T: Eq` where the
+   length of the tuple is above 12. The implementations of the standard
+   traits `PartialEq`, `Eq`, and `Default` are implemented in a macro
+   which is repeated for the lengths 1 - 12. See [src/libcore/tuple.rs]
+   for details.
+ - Working with function which have a variable number of arguments is
+   currently impossible. For example, a generic `bind` method,
+   `f.bind(a, b)(c) == f(a, b, c)`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Variadic generic parameters always take the form `...$ty`.
+
+### Defining types
+[defining-types]: #defining-types
+
+A `...T` in generic formal type parameters, where `...T` is the trailing
+type will define the type as variadic.
+
+```rust
+struct S<...T>(...T);
+
+S();
+S(42);
+S(12, 13, 17)
+```
+
+### Defining functions
+[defining-functions]: #defining-functions
+
+Similar to defining types the same syntax may be used in a functions generic
+parameters to define a variable length list of types.
+
+```rust
+fn bar<...T>(/* arguments */);
+```
+
+### Capturing multiple arguments
+[capturing-multiple-arguments]: #capturing-multiple-arguments
+
+In the functions arguments the pattern `...args: T` (where `T` is the name of
+the variadic list of types and `args` is the argument name) will be allowed
+to define a function which takes a variable number of arguments at the position
+of `args`.
+
+```rust
+fn bar<...T>(...args: T) -> R {...}
+
+bar();
+bar(42);
+bar(12, 13, 17);
+
+fn baz<...T>(n: i32, ...args: T) -> R { ... }
+
+baz(0);
+baz(1, 42);
+```
+
+### Unpacking variadic arguments
+[unpacking-variadic-arguments]: unpacking-variadic-arguments
+
+In order to unpack the variable length list of arguments in the function
+body, the pattern `...args` may be used in the function body.
+
+```rust
+fn f<...T>(...args: T) -> R {
+    // Unpack args and pass them to the function g
+    g(...args)
+}
+```
+
+### Defining bounds for `T`
+[defining-bounds-for-t]: #defining-bounds-for-t
+
+A type bound for each type in the variadic list of types `...T` may be
+defines with the syntax `...T: Trait`. This indicates that each type type
+in `T` must satisfy the bound `Trait`.
+
+```rust
+fn example<...T: Clone>(...args: T);
+```
+
+Now that the types in `T` can be bound and the list of arguments may be unpacked
+it becomes possible to iterate over the list of arguments and call trait methods
+on each argument in the list.
+
+```rust
+fn example<...T: Clone>(...args: T) {
+    // All types in T implement clone, so we can use
+    // the clone method on the unpacked args
+    f(...args.clone());
+}
+```
+
```

Comment:
The thing you wanted to solve (traits & impls) have not been demonstrated at all here... How do we abstract over arbitrary length tuples in impls and how do we do type level mapping (e.g. `(A, B)` to `(A::X, B::X)`) and things like http://hackage.haskell.org/package/type-list-0.5.0.0/docs/Data-Type-List.html? Also, how do we generalize https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.zip?